### PR TITLE
Flink-7035 Automatically identify AWS Region, simplified constructors, added test properties support

### PR DESCRIPTION
**Pull request by mpouttuclarke**

### CODE REVIEW

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java [hunk 1]
```diff
@@ -171,7 +172,14 @@ public static void validateAwsConfiguration(Properties config) {
 		}
 
 		if (!config.containsKey(AWSConfigConstants.AWS_REGION)) {
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java]: I wonder if its actually sufficient to rely on this check, without having the need to have a bunch of the new constructors that you added.

I.e., it is sufficient that the user doesn't supply a `AWSConfigConstants.AWS_REGION` value in the props, which we would then default to using `Regions.getCurrentRegion()`.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java [hunk 2]
```diff
@@ -171,7 +172,14 @@ public static void validateAwsConfiguration(Properties config) {
...
-			throw new IllegalArgumentException("The AWS region ('" + AWSConfigConstants.AWS_REGION + "') must be set in the config.");
+			final Region currentRegion = Regions.getCurrentRegion();
+			if (currentRegion != null) {
+				config.setProperty(AWSConfigConstants.AWS_REGION, currentRegion.getName());
+			} else {
+				throw new IllegalArgumentException("The AWS region could not be identified automatically from the AWS API.  " +
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java]: I would rephrase this:

>The AWS region could not be identified automatically from the environment.  Automatic AWS region retrieval is only available when the consumer is run in AWS environments, which is recommended for production usage to avoid cross-region network traffic. When executed in non-AWS environments, please set the region manually using the property AWSConfigConstants.AWS_REGION.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java (hunk 1 above)
Comment (mpouttuclarke) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java]: The new constructors make the easy path the right path.  We go through a lot of trouble at Amazon to make sure that the default constructors do the right thing with the minimal amount of effort.  Yet people still set things like region and auth manually when it is not only unnecessary but also a security, performance, and compliance risk.  Wherever we can we should try to follow the example of the AWS SDK and provide for using it correctly.  Overall, I would make the argument that using property files and statics isn't a best practice.  There really should be type safe POJOs and dependency injection in place for configuration of the consumer but that is a larger issue than I can take on right now.  The new constructors attempt to add some type safety while improving ease of use when operating in an Amazon environment.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java (hunk 2 above)
Comment (mpouttuclarke) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/util/KinesisConfigUtil.java]: Yes I like that wording.

File: docs/dev/connectors/kinesis.md [hunk 3]
```diff
@@ -72,12 +72,80 @@ Before consuming data from Kinesis streams, make sure that all streams are creat
...
+val kinesis = env.addSource(new FlinkKinesisConsumer[String](
+    "kinesis_stream_name", new SimpleStringSchema))
+{% endhighlight %}
+</div>
+</div>
+
+Configuration for the consumer can also be supplied with `java.util.Properties` for use on non-Amazon Linux hardware,
+or in the case that other stream consumer properties need to be tuned.
+
+Please note it is strongly recommended to use Kinesis streams within the same availability zone they originate in.
```
Comment (mpouttuclarke) [inline docs/dev/connectors/kinesis.md]: We completely restrict cross-region network traffic except in special circumstances within Amazon because of these reasons.  These are simply lessons learned from scaling our systems globally, where situations arise not only due to performance concerns but also regulatory issues such as EU data laws restricting data egress for example.  Customer should be aware of these issues and discuss them with support before going down this path.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java [hunk 4]
```diff
@@ -117,12 +117,77 @@
...
+	 * @param initialPosition
+	 *           Where to start the Kinesis stream
+	 */
+	public FlinkKinesisConsumer(String stream, KinesisDeserializationSchema<T> deserializer, InitialPosition initialPosition) {
+		this(stream, deserializer, getDefaultPropsWithInitialPosition(initialPosition));
+	}
+
+	private static Properties getDefaultPropsWithInitialPosition(InitialPosition initialPosition) {
+		final Properties conf = getDefaultConfigProperties();
+		conf.setProperty(ConsumerConfigConstants.STREAM_INITIAL_POSITION, initialPosition.name());
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java]: This properties building is problematic. For example, we also have an `InitialPosition.AT_TIMESTAMP` configuration, which involves 2 keys in the properties.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java [hunk 5]
```diff
@@ -117,12 +117,77 @@
 	// ------------------------------------------------------------------------
 	//  Constructors
 	// ------------------------------------------------------------------------
+	/**
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java]: nit: add empty line before /**

File: docs/dev/connectors/kinesis.md (hunk 3 above)
Comment (tzulitai) [inline docs/dev/connectors/kinesis.md]: Thanks for the explanation! That makes a lot of sense.

File: flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java [hunk 6]
```diff
@@ -117,12 +117,77 @@
...
+		this(stream, new KinesisDeserializationSchemaWrapper<>(deserializer));
+	}
+
+	/**
+	 * Creates a new Flink Kinesis Consumer that uses the AWS credentials and region provided from the AWS
+	 * node with the provided initial position in the stream.
+	 *
+	 * @param stream
+	 * @param deserializer
+	 * @param initialPosition
```
Comment (tzulitai) [inline flink-connectors/flink-connector-kinesis/src/main/java/org/apache/flink/streaming/connectors/kinesis/FlinkKinesisConsumer.java]: These params need descriptions for the Javadoc also.

**Comment by tzulitai** [comment]
I'm taking over the remaining issues of this pull request.

**Comment by tzulitai** [comment]
For my additional fixes, I would remove the additional constructors for now, since the configuration parsing issue is yet to be discussed and addressed. Will open another JIRA for that.

**Comment by aljoscha** [comment]
@tweise Do you think this one is still relevant? You guys must have the most Kinesis experience.

**Comment by aljoscha** [comment]
I'm closing this as "Abandoned", since there is no more activity and the code base has moved on quite a bit. Please re-open this if you feel otherwise and work should continue.
